### PR TITLE
Parameter assignment syntax

### PR DESCRIPTION
**Pull request by jemc**
This PR adds my RFC for assign param syntax.

### CODE REVIEW

**Comment by aturley** [comment]
Discuss more next week.

File: text/0000-assign-param-syntax.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+# How We Test This
+
+New compiler tests that demonstrate this case, including the edge cases mentioned in the implementation details above.
+
+# Drawbacks
+
+* New patterns to learn (though no codebases will be required to use these patterns, you have to learn the new patterns if you want to write or read code that uses them).
+* Added compiler complexity.
+
```
Comment (jasoncarr0) [inline text/0000-assign-param-syntax.md:106]: Symbols/glyphs are hard to understand and search for, so this makes code less explicit and thus possibly harder to learn, or more intimidating

Comment (jemc) [reply text/0000-assign-param-syntax.md:106]: I personally don't really buy this argument in this case, because if we used the `@` symbol for `this.` (and if `this.` were mandatory), as proposed in the other RFC, this symbol would be used in nearly every function of every file of every Pony program.

Sure, somebody who was learning the language for the first time won't immediately know what it means, but they'd learn what it means within the first 5-10 minutes of learning, and they wouldn't forget because it would be so ubiquitous. Searchability of the symbol in search engines is not really relevant, because anyone encountering it for the first time would also be learning the language for the first time, so they'd already be looking at the tutorial.

I actually think that your argument is a much more compelling argument *in favor* of RFC #173 instead of against it - that is, it is a compelling argument for us to *not* use `@` for FFI or bare functions (which are rare/arcane), and to use it for the much more ubiquitous `this.` instead. FFI code is much more obscure, and a user could be using Pony for years before encountering their first reading of FFI code. Hence, if and when they do encounter the `@` symbol for FFI or a bare function they are likely to be confused, and find it hard to search for.

Note that in the Savi language I followed this design principle - it uses the `@` symbol for self member access, but FFI usage and declaration has no special symbol - FFI looks like a normal type declaration at the declaration site, and a normal call at the use site - it just happens to be a type declaration that is marked as being `:ffi` rather than some other kind of type. In this regime, users can quickly adapt to understanding `@`, but when they get to the more obscure concepts like FFI there are no new symbols to learn / search fo.

**Comment by SeanTAllen** [comment]
@jemc how would you feel about moving forward with `this.` instead of `@`? I'd like to move this and #173 forward but we have some syntax issues to resolve. I'm in favor of "this" because it is the least breaking option.

**Comment by jemc** [comment]
I'd agree that `this.` is "better than nothing" for the purposes of this RFC, so I'd be good with amending this RFC to use that syntax. It's still less boilerplate than the current status quo.

For #173 I would not be in favor of making that change with `this.` - I think it's too much friction added for something so ubiquitous.

**Comment by SeanTAllen** [comment]
We appear to have an impasse on #173, but it would be good to move forward with this RFC so let's do that and continue to discuss #173.

File: text/0000-assign-param-syntax.md (hunk 1 above)
Comment (SeanTAllen) [reply text/0000-assign-param-syntax.md:106]: This now has `this.` instead of `@` so this is resolved for now.

File: text/0000-assign-param-syntax.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  let name: (Id | LitString)
+  let return_type: TypeArgs
+  let params: Params
+  let partial: (Question | None)
+  let guard: (IfDefCond | None)
+
+  new val create(name: String, this.return_type, this.params, this.attachments = None) =>
+    this.name = Id(name)
+    this.partial = None
+    this.guard = None
```
Comment (jasoncarr0) [inline text/0000-assign-param-syntax.md:77]: What happens with this PR if the body of the constructor is now empty? Do we still need to include `=> None`?

Comment (SeanTAllen) [reply text/0000-assign-param-syntax.md:77]: My expectation would be yes, you need to. I think it is worth explicitly noting that in this RFC.

File: text/0000-assign-param-syntax.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+```pony
+class val UseFFIDecl is (AST & UseDecl)
+  let attachments: (Attachments | None)
+  let name: (Id | LitString)
+  let return_type: TypeArgs
+  let params: Params
+  let partial: (Question | None)
+  let guard: (IfDefCond | None)
+
+  new val create(this.name, this.return_type, this.params, this.partial, this.guard, this.attachments)
```
Comment (SeanTAllen) [inline text/0000-assign-param-syntax.md]: this should have `=> None` yes?

**Comment by SeanTAllen** [comment]
Is there an implicit consume of `iso` parameters?

File: text/0000-assign-param-syntax.md (hunk 2 above)
Comment (jemc) [reply text/0000-assign-param-syntax.md:77]: I've updated it to explicitly mention that requirement.

File: text/0000-assign-param-syntax.md [hunk 4]
```diff
@@ -55,10 +55,10 @@ class val UseFFIDecl is (AST & UseDecl)
...
   let guard: (IfDefCond | None)
 
-  new val create(this.name, this.return_type, this.params, this.partial, this.guard, this.attachments)
+  new val create(this.name, this.return_type, this.params, this.partial, this.guard, this.attachments) => None
 ```
 
-Note that there is now no syntactical body of the method, but the compiler will add a body that does the appropriate assignments.
+Note that there is now no syntactical body of the method apart from the obligatory `None` expression, and the compiler will add a body that does the appropriate assignments.
 
 Of course, "assign parameters" can be used alongside normal parameters, and not every field need be associated to an "assign parameter". Assign parameters can also have a "default value" as normal parameters can, which specifies the value to use if the caller does not supply an argument for that position. The following refactored example demonstrates all of those concepts:
```
Comment (jasoncarr0) [inline text/0000-assign-param-syntax.md:63]: One more note possibly worth adding: are any names supported for keyword argument syntax?

**Comment by jasoncarr0** [comment]
One more note possibly worth adding: are any names supported for keyword argument syntax? Does that create semver issues with the name requirements?

**Comment by SeanTAllen** [comment]
It was noted on the sync call that @jemc will note that how this interacts with named arguments in unresolved questions.

**Review by jasoncarr0 — APPROVED** [review]

**Comment by SeanTAllen** [comment]
This was accepted during sync. I will do the merge and what not later.
